(#52) Shared CI workflow for APIs.

blairlearn opened the pull request:
This is the foundational commit for having a single workflow file among the APIs.

These PRs depend on this one being merged and released.

- NCIOCPL/glossary-api#177
- NCIOCPL/drug-dictionary-api#95
- NCIOCPL/sitewide-search-api#110
- NCIOCPL/bestbets-api#61
- NCIOCPL/r4r-api#29
- NCIOCPL/clinical-trials-listing-api#66

Closes #52
Closes #56

jfrank-nih reviewed (commented):
- Probably worth adding the workflow concurrency per API project/PR
- We have taken to uploading backstop js reports to Akamai to give us an easy viewing location. We might want to make a generic net storage group for such things and then we could send integration test results there too. It's Nice to just click on a link in the PR if the tests have failed rather than downloading the artifacts.

**Very Minor:**
- Is matrix concurrency for macOS needed? It doesn't hurt, but realistically we're never going to run these on Mac
- Is there a reason we're not saving code coverage outputs for windows?

blairlearn commented:
@jfrank-nih 

- Workflow concurrency: Totally worthwhile.  It's added.
- Uploading integration results: I'm going to hold off on that. Integration tests can (and should!) be run locally before submitting a PR.
- Running unit tests on Mac: As you said, it doesn't hurt, and we do our development on Mac, so this is an extra check to make sure they run there.
- Not saving code coverage outputs on Windows: For whatever reason, on Windows, the coverage tests get confused about the number of test projects being specified (only one exists) and the tests fail.  I've added a comment to explain why Windows is being treated differently.

blairlearn committed:
(#52, #56) Shared CI workflow for APIs

- Add concurrency limit to main workflow.
- Clean up handling of integration test failure.
- Bump *-artifact actions to @v4
  - Fix problem with coverage file being uploaded multiple times with
    the same name.
- Bump action versions.
- Introduce global.json to set the invoked SDK version.
- Move deploy-api.ps1 (formerly deploy-site.ps1) to the central repository.

Closes #52
Closes #56

jfrank-nih commented on .github/workflows/common-api-ci.yml:
```diff
@@ -0,0 +1,201 @@
+on:
+  workflow_call:
+    inputs:
+      # Path to the main API project.
```
`workflow_call` can still have a `description` key under it, even though it's not visible anywhere. I'd put your comments into that for clarity.

blairlearn replied:
Fair.  And it's not too hard to imagine tooling getting smart enough to pull up the description when writing the calling workflow.

jfrank-nih commented on .github/workflows/common-api-ci.yml:
```diff
@@ -0,0 +1,201 @@
+        type: string
+
+# Only run one instance of this workflow at a time per branch.
+concurrency:
```
I'm pretty sure that `concurrency` checks need to go in the individual repos, not in the shared one.

blairlearn replied:
Tested by pushing two changes a short time apart. The earlier one ended up being canceled with the message "Canceling since a higher priority waiting request for '177/merge-Glossary API CI workflow' exists."

Which is nice as it's that much less boiler-plate in the calling workflow.  (I can't imagine a situation where we'd actually want the same workflow running multiple times for a  single branch.)

![image](https://github.com/user-attachments/assets/2457fa9c-4067-420c-8435-cc347bae5880)

jfrank-nih commented on .github/workflows/common-api-ci.yml:
```diff
@@ -0,0 +1,201 @@
+              else
+                dotnet test /p:CollectCoverage=true /p:CoverletOutputFormat=lcov /p:CoverletOutput=../../lcov test/**
+              fi
+        shell: bash
```
Minor, but I'd move this before the `run` just for readability.

blairlearn replied:
While I was at it, I made some of the implicit `shell: bash` instances explicit.

jfrank-nih commented on .github/workflows/common-api-ci.yml:
```diff
@@ -0,0 +1,201 @@
+            # Copy deploy toool.
+            cp $GITHUB_WORKSPACE/shared/tools/deploy-api.ps1 $GITHUB_WORKSPACE/out/
+
+        shell: bash
```
Same as above.

jfrank-nih commented on .github/workflows/common-api-ci.yml:
```diff
@@ -0,0 +1,201 @@
+
+      - name: Run Integration Test
+        id: integration_tests
+        continue-on-error: true
```
I think that `continue-on-error` [can mess with whether your job](https://stackoverflow.com/questions/62045967/is-there-a-way-to-continue-on-error-while-still-getting-correct-feedback#comment137995684_73357322) is marked as `failed` or not. It might also make the step outcome always be `success`? 

I'd suggest removing it, marking the upload step below as `if: ${{ always() }}` and removing the last step.

blairlearn replied:
I think this is from an earlier replacement for an even earlier workaround from before GHA had the `if: always()` support.  But yes, I agree, using `always()` is a good deal cleaner.

jfrank-nih replied:
You don't want the `continue-on-error` anymore now, because `always()` will fire anyway.

jfrank-nih commented on .github/workflows/common-api-ci.yml:
```diff
@@ -0,0 +1,201 @@
+        shell: bash
+
+      - name: Record metadata
+        uses: nciocpl/build-info-action@v1.0.0
```
Should this be v1.0.0? (really, we should probably do @v1, but that requires improving the tooling on that action to handle semver tags)

blairlearn replied:
Whoops!  No, that should indeed be v1.0.1.

(And there's a lot that could be done to improve that action and/or make it more useful.)

jfrank-nih commented on .github/workflows/workflow.yml:
```diff
@@ -77,7 +83,7 @@ jobs:
 
       - name: Upload Published Artifact
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
```
Can't comment on the build-info-action line (5 above), so just noting here that it should probably be v1.0.1

jfrank-nih commented on .github/workflows/workflow.yml:
`test_build_release` is probably similar enough to the same job in `common-api-ci.yml` that you _could_ point this to that, with a few input modifications. But I don't think `integration_tests` is similar enough. So that would require splitting the common workflow in half so you could use just part of it. In the end, probably not worth it for the re-use.

jfrank-nih replied:
Is it worth bringing this up to elasticsearch 7.17 so we're using the same version all around? Or not really relevant?

blairlearn replied:
Yes.  Worth updating.  (This is the first time the nuget packages have been updated since .Net 3.)

blairlearn replied:
`test_build_release` - they're _similar_.  But the one in `workflow.yml` build multiple packages, which would require parameterization.  It also uses a different version of `.Net` (that will be addressed in #54).

jfrank-nih commented on tools/deploy-api.ps1:
```diff
@@ -0,0 +1,255 @@
+	.PARAMETER app_version
+	The version of the API for the application. Required.
+	.PARAMETER site_hostname
+	The hostname of te API website. Required.
```
`te` > `the`

blairlearn replied:
OK, this one I'll fix. 🙂

But for the rest... This is an existing script; it's inclusion here is intended to  eliminate the need for each API repo to maintain a separate copy.

Refactoring is beyond the scope of issue #53.  That cleanup is [issue 57](https://github.com/NCIOCPL/NCI.OCPL.Api.Shared/issues/57)

jfrank-nih replied:
It shows as a new file in this PR for whatever reason.

jfrank-nih commented on tools/deploy-api.ps1:
```diff
@@ -0,0 +1,255 @@
+
+function CreateSite() {
+	## Check parameters
+	if ($site_name -eq "")
```
Should we have a `function CheckParam([string]$param, [string]$ParamName)` to do these checks?

jfrank-nih commented on tools/deploy-api.ps1:
```diff
@@ -0,0 +1,255 @@
+	}
+}
+
+function Is-Admin {
```
Any particular reason this one function name is hyphenated?

jfrank-nih commented on tools/deploy-api.ps1:
```diff
@@ -0,0 +1,255 @@
+		Stop-Website -Name $site_name
+	}
+	else {
+		write-host "*********** False *************";
```
I can't see the overall program output but it seems like this `write-host` line doesn't have enough info or context. Maybe "No website exists to stop" or something like that?

bryanpizzillo commented on .github/workflows/common-api-ci.yml:
```diff
@@ -0,0 +1,198 @@
+              cd integration-tests && ./bin/karate ./features
+
+      - name: Upload Integration test results
+        if: always()
```
Is this the correct syntax here? The docs say it should be `if: ${{ always() }}`
https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#always

blairlearn replied:
The curly brackets are optional on the `if` ([Docs](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idif))

> When you use expressions in an if conditional, you can, optionally, omit the ${{ }} expression syntax because GitHub Actions automatically evaluates the if conditional as an expression. However, this exception does not apply everywhere.
